### DISCUSSION TITLE
关于sock.htonl()

### DISCUSSION THREAD

**Discussion by lxr12010513** [SUSTech-CS305-Fall22/CS305-Project-Skeleton#22] [Q&A]
example中struct.pack有时候用sock.htonl(Seq)，有时候又直接用Seq, 请问socket.htonl()的作用是什么?

**Comment by TANSixu** [comment]
Hi,感谢提问！

htonl()的作用是将4byte的主机字节序(host)数据转成网络字节序(network)数据。
在我们需要把一个int类型的整数（在python里就是超过2byte short范围的其他整数）用struct放进报文里的时候，我们需要用htonl（host to network long）。在我们需要把网络报文中的4byte转成python的int整数的时候，我们需要ntohl（network to host long）。

那为什么有的情况直接放的Seq呢？非常抱歉，其实是因为我的命名写的不好。看一下出现直接放Seq的地方：

```
def process_inbound_udp(sock):
    # Receive pkt
    pkt, from_addr = sock.recvfrom(BUF_SIZE)
    Magic, Team, Type,hlen, plen, Seq, Ack= struct.unpack("HBBHHII", pkt[:HEADER_LEN])
    data = pkt[HEADER_LEN:]

.......

    elif Type == 3:
        # received a DATA pkt
        ex_received_chunk[ex_downloading_chunkhash] += data

        # send back ACK
        ack_pkt = struct.pack("HBBHHII", socket.htons(52305),35,  4,socket.htons(HEADER_LEN), socket.htons(HEADER_LEN), 0, Seq)
        sock.sendto(ack_pkt, from_addr)


........
```
出现在dumbreceiver.py的process_inbound_udp中的创建ACK报文的地方：
```
ack_pkt = struct.pack("HBBHHII", socket.htons(52305),35,  4,socket.htons(HEADER_LEN), socket.htons(HEADER_LEN), 0, Seq)
```
需要注意的是，这里的Seq变量是从这个方法开头
```
Magic, Team, Type,hlen, plen, Seq, Ack= struct.unpack("HBBHHII", pkt[:HEADER_LEN])
```
直接从网络包里读出来的，并且它还没有经过ntohl的转换（加入你的序号本来应该是11，这个时候这个Seq的值并不是11而是一个很大的整数，因为它还是网络字节序），因此它现在本来就**还是网络字节序**，直接放进报文即可。

这个命名实在不好，有些误导，应该改为
```
Magic_raw, Team, Type,hlen_raw, plen_raw, Seq_raw, Ack_raw= struct.unpack("HBBHHII", pkt[:HEADER_LEN])
```
注意Type和Team不需要转换，因为他俩只有一个字节，不需要考虑字节序。

非常抱歉！

如果还是比较疑惑的话，请再与我们讨论！

**Reply by ZephyrusZhang** [reply]
请问如果`htonl()`是为了转换成网络字节序的话，直接将struct format中的byte order指定为`!`能不能达到相同目的呢？见官方文档：[Byte Order, Size, and Alignment](https://docs.python.org/3/library/struct.html#byte-order-size-and-alignment)。

**Reply by TANSixu** [reply]
> 请问如果`htonl()`是为了转换成网络字节序的话，直接将struct format中的byte order指定为`!`能不能达到相同目的呢？见官方文档：[Byte Order, Size, and Alignment](https://docs.python.org/3/library/struct.html#byte-order-size-and-alignment)。

hi 可以的可以的，这比htonl方便多了，可惜之前我没有注意到这个特性！非常感谢提出。

不管是pack还是unpack，直接加“！”就可以，会自动转换。

再次感谢！
